nni/nas/oneshot/pytorch/strategy.py

```diff
@@ -12,223 +12,951 @@
+        return {
+            'train': train_dataloader_fn(),
+            'val': val_dataloader_fn()
+        }
+
+    def val_dataloader(self, train_dataloader_fn, val_dataloader_fn):
+        return None
+
+
+class Proxyless(DARTS):
+    __doc__ = """
+    A low-memory-consuming optimized version of differentiable architecture search. See `reference <https://arxiv.org/abs/1812.00332>`__.
+
+    This is a :class:`~nni.nas.strategy.DARTS`-based method that resamples the architecture to reduce memory consumption.
+    Essentially, it samples one path on forward,
+    and implements its own backward to update the architecture parameters based on only one path.
+
+    The supported mutation primitives of :class:`Proxyless` are:
+
+    * :class:`nni.nas.nn.pytorch.LayerChoice` (candidate layers must NOT have keyword arguments).
+    * :class:`nni.nas.nn.pytorch.InputChoice`.
+    * :class:`nni.nas.nn.pytorch.Repeat` (with categorical choice of no transformation).
+
+    {optimization_note}
+
+    Parameters
+    ----------
+    **kwargs
+        Supported parameters are the same as :class:`~nni.nas.strategy.DARTS`.
+    """.format(
+        optimization_note=MANUAL_OPTIMIZATION_NOTE
+    )
+
+    def default_mutation_hooks(self) -> list[MutationHook]:
+        """Replace modules with proxyless-differentiable versions."""
+        from .supermodule.proxyless import ProxylessMixedLayer, ProxylessMixedInput, ProxylessMixedRepeat, suppress_already_mutated
+
+        hooks = [
+            suppress_already_mutated,
+            ProxylessMixedLayer.mutate,
+            ProxylessMixedInput.mutate,
+            ProxylessMixedRepeat.mutate,
+            no_default_hook,
+        ]
+        # FIXME: no support for mixed operation currently
+        return hooks
+
+
+class GumbelDARTS(DARTS):
+    __doc__ = """
+    Choose the best block by using Gumbel Softmax random sampling and differentiable training.
+    See `FBNet <https://arxiv.org/abs/1812.03443>`__ and `SNAS <https://arxiv.org/abs/1812.09926>`__.
+
+    This is a :class:`~nni.nas.strategy.DARTS`-based method that uses gumbel-softmax to simulate one-hot distribution.
+    Essentially, it tries to mimick the behavior of sampling one path on forward by gradually
+    cool down the temperature, aiming to bridge the gap between differentiable architecture weights and
+    discretization of architectures.
+
+    .. versionadded:: 2.8
+    
+       Supports searching for ValueChoices on operations, with the technique described in
+       `FBNetV2: Differentiable Neural Architecture Search for Spatial and Channel Dimensions <https://arxiv.org/abs/2004.05565>`__.
+
+    The supported mutation primitives of GumbelDARTS are:
+
+    * :class:`nni.nas.nn.pytorch.LayerChoice`.
+    * :class:`nni.nas.nn.pytorch.InputChoice`.
+    * :class:`nni.nas.nn.pytorch.ParameterizedChoice` (only when parameters are choices and type is in {supported_ops}).
```

Comment:
what is the difference between ParameterizedChoice and ValueChoice?

Reply:
Typo here. It should be `ParametrizedModule`, which is a `nn.Module` with ValueChoice in its init argument.